### ISSUE TITLE
greedy upgrade repeatedly downloads and reinstall same version

### ISSUE THREAD

**Issue by reagle** [caskroom/homebrew-cask#45628]
#### Description of issue

I have three apps (dropbox latest, google-backup-and-sync latest, pdftotext latest) that **always** download and upgrade themselves whenever I run `brew cask upgrade --greedy`.

#### Output of your command with `--verbose --debug`

```
╭─reagle@hom ~
╰─➤  brew cask upgrade --greedy --verbose --debug
==> Upgrading 3 outdated packages, with result:
dropbox latest, google-backup-and-sync latest, pdftotext latest
==> Started upgrade process for Cask dropbox
==> Hbc::Installer#fetch
==> Satisfying dependencies
==> Downloading
==> Downloading https://www.dropbox.com/download?plat=mac&full=1
/usr/bin/curl --show-error --user-agent Homebrew/1.5.14-3-g94c0d83 (Macintosh; Intel Mac OS X 10.13.3) curl/7.54.0 --fail --location --remote-time --continue-at - --output /Users/reagle/Library/Caches/Homebrew/Cask/dropbox--latest.incomplete https://www.dropbox.com/download?plat=mac&full=1
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100  112M  100  112M    0     0  7993k      0  0:00:14  0:00:14 --:--:-- 7088k
==> Downloaded to -> /Users/reagle/Library/Caches/Homebrew/Cask/dropbox--latest
==> Verifying download
==> Determining which verifications to run for Cask dropbox
==> Checking for verification class Hbc::Verify::Checksum
==> No checksum defined for Cask dropbox, skipping verification
==> 0 verifications defined
==> Starting upgrade for Cask dropbox
==> Un-installing artifacts
==> 3 artifact/s defined
#<SortedSet:0x007f80ab439df0>
==> Un-installing artifact of class Hbc::Artifact::Uninstall
==> Running uninstall process for dropbox; your password may be necessary
==> Removing launchctl service com.dropbox.DropboxMacUpdate.agent
==> Executing: ["/bin/launchctl", "list", "com.dropbox.DropboxMacUpdate.agent"]
==> Executing: ["/usr/bin/sudo", "-E", "--", "/bin/launchctl", "list", "com.dropbox.DropboxMacUpdate.agent"]
==> Removing login item Dropbox
==> Executing: ["/usr/bin/osascript", "-e", "tell application \"System Events\" to delete every login item whose name is \"Dropbox\""]
==> Un-installing artifact of class Hbc::Artifact::App
==> Backing App 'Dropbox.app' up to '/usr/local/Caskroom/dropbox/latest/Dropbox.app'.
==> Removing App '/Applications/Dropbox.app'.
==> Hbc::Installer#stage
==> Extracting primary container
==> Determining which containers to use based on filetype
==> Checking container class Hbc::Container::Pkg
==> Checking container class Hbc::Container::Ttf
==> Checking container class Hbc::Container::Otf
==> Checking container class Hbc::Container::Air
==> Checking container class Hbc::Container::Cab
==> Checking container class Hbc::Container::Dmg
==> Executing: ["/usr/bin/hdiutil", "imageinfo", "/Users/reagle/Library/Caches/Homebrew/Cask/dropbox--latest"]
==> Using container class Hbc::Container::Dmg for /Users/reagle/Library/Caches/Homebrew/Cask/dropbox--latest
==> Executing: ["/usr/bin/hdiutil", "attach", "-plist", "-nobrowse", "-readonly", "-noidme", "-mountrandom", "/var/folders/2l/73vdx0sd5rvcn38yg036h6500000gn/T/d20180405-30789-cnqs40", "/Users/reagle/Library/Caches/Homebrew/Cask/dropbox--latest"]
==> Executing: ["/usr/bin/find", ".", "-print0"]
==> Executing: ["/usr/bin/mkbom", "-s", "-i", "/var/folders/2l/73vdx0sd5rvcn38yg036h6500000gn/T/20180405-30789-1tpzbf9.list", "--", "/var/folders/2l/73vdx0sd5rvcn38yg036h6500000gn/T/20180405-30789-1jbl34j.bom"]
==> Executing: ["/usr/bin/ditto", "--bom", "/var/folders/2l/73vdx0sd5rvcn38yg036h6500000gn/T/20180405-30789-1jbl34j.bom", "--", "/private/var/folders/2l/73vdx0sd5rvcn38yg036h6500000gn/T/d20180405-30789-cnqs40/dmg.BNqOFp", "/usr/local/Caskroom/dropbox/latest"]
==> Executing: ["/usr/sbin/diskutil", "eject", "/private/var/folders/2l/73vdx0sd5rvcn38yg036h6500000gn/T/d20180405-30789-cnqs40/dmg.BNqOFp"]
==> Creating metadata directory /usr/local/Caskroom/dropbox/.metadata/latest/20180405203708.617.
==> Creating metadata subdirectory /usr/local/Caskroom/dropbox/.metadata/latest/20180405203708.617/Casks.
==> Installing artifacts
==> 3 artifact/s defined
#<SortedSet:0x007f80ab431768>
==> Installing artifact of class Hbc::Artifact::App
==> Moving App 'Dropbox.app' to '/Applications/Dropbox.app'.
==> Purging files for version latest of Cask dropbox
🍺  dropbox was successfully upgraded!
==> Started upgrade process for Cask google-backup-and-sync
==> Hbc::Installer#fetch
==> Satisfying dependencies
==> Downloading
==> Downloading https://dl.google.com/drive/InstallBackupAndSync.dmg
/usr/bin/curl --show-error --user-agent Homebrew/1.5.14-3-g94c0d83 (Macintosh; Intel Mac OS X 10.13.3) curl/7.54.0 --fail --location --remote-time --continue-at - --output /Users/reagle/Library/Caches/Homebrew/Cask/google-backup-and-sync--latest.dmg.incomplete https://dl.google.com/drive/InstallBackupAndSync.dmg
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 42.7M  100 42.7M    0     0  7885k      0  0:00:05  0:00:05 --:--:-- 8621k
==> Downloaded to -> /Users/reagle/Library/Caches/Homebrew/Cask/google-backup-and-sync--latest.dmg
==> Verifying download
==> Determining which verifications to run for Cask google-backup-and-sync
==> Checking for verification class Hbc::Verify::Checksum
==> No checksum defined for Cask google-backup-and-sync, skipping verification
==> 0 verifications defined
==> Starting upgrade for Cask google-backup-and-sync
==> Un-installing artifacts
==> 3 artifact/s defined
#<SortedSet:0x007f80ab4b2908>
==> Un-installing artifact of class Hbc::Artifact::Uninstall
==> Running uninstall process for google-backup-and-sync; your password may be necessary
==> Quitting application ID com.google.GoogleDrive
==> Executing: ["/bin/launchctl", "list"]
==> Removing login item Backup and sync from Google
==> Executing: ["/usr/bin/osascript", "-e", "tell application \"System Events\" to delete every login item whose name is \"Backup and sync from Google\""]
==> Un-installing artifact of class Hbc::Artifact::App
==> Backing App 'Backup and Sync.app' up to '/usr/local/Caskroom/google-backup-and-sync/latest/Backup and Sync.app'.
==> Removing App '/Applications/Backup and Sync.app'.
==> Hbc::Installer#stage
==> Extracting primary container
==> Determining which containers to use based on filetype
==> Checking container class Hbc::Container::Pkg
==> Checking container class Hbc::Container::Ttf
==> Checking container class Hbc::Container::Otf
==> Checking container class Hbc::Container::Air
==> Checking container class Hbc::Container::Cab
==> Checking container class Hbc::Container::Dmg
==> Executing: ["/usr/bin/hdiutil", "imageinfo", "/Users/reagle/Library/Caches/Homebrew/Cask/google-backup-and-sync--latest.dmg"]
==> Using container class Hbc::Container::Dmg for /Users/reagle/Library/Caches/Homebrew/Cask/google-backup-and-sync--latest.dmg
==> Executing: ["/usr/bin/hdiutil", "attach", "-plist", "-nobrowse", "-readonly", "-noidme", "-mountrandom", "/var/folders/2l/73vdx0sd5rvcn38yg036h6500000gn/T/d20180405-30789-120glcg", "/Users/reagle/Library/Caches/Homebrew/Cask/google-backup-and-sync--latest.dmg"]
==> Executing: ["/usr/bin/find", ".", "-print0"]
==> Executing: ["/usr/bin/mkbom", "-s", "-i", "/var/folders/2l/73vdx0sd5rvcn38yg036h6500000gn/T/20180405-30789-va587x.list", "--", "/var/folders/2l/73vdx0sd5rvcn38yg036h6500000gn/T/20180405-30789-1f15int.bom"]
==> Executing: ["/usr/bin/ditto", "--bom", "/var/folders/2l/73vdx0sd5rvcn38yg036h6500000gn/T/20180405-30789-1f15int.bom", "--", "/private/var/folders/2l/73vdx0sd5rvcn38yg036h6500000gn/T/d20180405-30789-120glcg/dmg.CUx4M3", "/usr/local/Caskroom/google-backup-and-sync/latest"]
==> Executing: ["/usr/sbin/diskutil", "eject", "/private/var/folders/2l/73vdx0sd5rvcn38yg036h6500000gn/T/d20180405-30789-120glcg/dmg.CUx4M3"]
==> Creating metadata directory /usr/local/Caskroom/google-backup-and-sync/.metadata/latest/20180405203730.985.
==> Creating metadata subdirectory /usr/local/Caskroom/google-backup-and-sync/.metadata/latest/20180405203730.985/Casks.
==> Installing artifacts
==> 3 artifact/s defined
#<SortedSet:0x007f80ab4a1e28>
==> Installing artifact of class Hbc::Artifact::App
==> Moving App 'Backup and Sync.app' to '/Applications/Backup and Sync.app'.
==> Purging files for version latest of Cask google-backup-and-sync
🍺  google-backup-and-sync was successfully upgraded!
==> Started upgrade process for Cask pdftotext
==> Hbc::Installer#fetch
==> Satisfying dependencies
==> Downloading
==> Downloading https://www.bluem.net/files/pdftotext.dmg
/usr/bin/curl --show-error --user-agent Homebrew/1.5.14-3-g94c0d83 (Macintosh; Intel Mac OS X 10.13.3) curl/7.54.0 --fail --location --remote-time --continue-at - --output /Users/reagle/Library/Caches/Homebrew/Cask/pdftotext--latest.dmg.incomplete https://www.bluem.net/files/pdftotext.dmg
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  829k  100  829k    0     0   734k      0  0:00:01  0:00:01 --:--:--  735k
==> Downloaded to -> /Users/reagle/Library/Caches/Homebrew/Cask/pdftotext--latest.dmg
==> Verifying download
==> Determining which verifications to run for Cask pdftotext
==> Checking for verification class Hbc::Verify::Checksum
==> No checksum defined for Cask pdftotext, skipping verification
==> 0 verifications defined
==> Starting upgrade for Cask pdftotext
==> Un-installing artifacts
==> 2 artifact/s defined
#<SortedSet:0x007f80ab482ca8>
==> Un-installing artifact of class Hbc::Artifact::Uninstall
==> Running uninstall process for pdftotext; your password may be necessary
==> Uninstalling packages:
==> Executing: ["/usr/sbin/pkgutil", "--pkgs=net.bluem.pdftotext.pkg"]
net.bluem.pdftotext.pkg
==> Executing: ["/usr/sbin/pkgutil", "--files", "net.bluem.pdftotext.pkg"]
==> Executing: ["/usr/sbin/pkgutil", "--pkg-info-plist", "net.bluem.pdftotext.pkg"]
==> Deleting pkg files
==> Executing: ["/usr/bin/sudo", "-E", "--", "/usr/bin/xargs", "-0", "--", "/bin/rm", "--"]
==> Unregistering pkg receipt (aka forgetting)
==> Executing: ["/usr/bin/sudo", "-E", "--", "/usr/sbin/pkgutil", "--forget", "net.bluem.pdftotext.pkg"]
==> Hbc::Installer#stage
==> Extracting primary container
==> Determining which containers to use based on filetype
==> Checking container class Hbc::Container::Pkg
==> Checking container class Hbc::Container::Ttf
==> Checking container class Hbc::Container::Otf
==> Checking container class Hbc::Container::Air
==> Checking container class Hbc::Container::Cab
==> Checking container class Hbc::Container::Dmg
==> Executing: ["/usr/bin/hdiutil", "imageinfo", "/Users/reagle/Library/Caches/Homebrew/Cask/pdftotext--latest.dmg"]
==> Using container class Hbc::Container::Dmg for /Users/reagle/Library/Caches/Homebrew/Cask/pdftotext--latest.dmg
==> Executing: ["/usr/bin/hdiutil", "attach", "-plist", "-nobrowse", "-readonly", "-noidme", "-mountrandom", "/var/folders/2l/73vdx0sd5rvcn38yg036h6500000gn/T/d20180405-30789-un18q8", "/Users/reagle/Library/Caches/Homebrew/Cask/pdftotext--latest.dmg"]
==> Executing: ["/usr/bin/find", ".", "-print0"]
==> Executing: ["/usr/bin/mkbom", "-s", "-i", "/var/folders/2l/73vdx0sd5rvcn38yg036h6500000gn/T/20180405-30789-40rfsd.list", "--", "/var/folders/2l/73vdx0sd5rvcn38yg036h6500000gn/T/20180405-30789-178p9nb.bom"]
==> Executing: ["/usr/bin/ditto", "--bom", "/var/folders/2l/73vdx0sd5rvcn38yg036h6500000gn/T/20180405-30789-178p9nb.bom", "--", "/private/var/folders/2l/73vdx0sd5rvcn38yg036h6500000gn/T/d20180405-30789-un18q8/dmg.DXLEyV", "/usr/local/Caskroom/pdftotext/latest"]
==> Executing: ["/usr/sbin/diskutil", "eject", "/private/var/folders/2l/73vdx0sd5rvcn38yg036h6500000gn/T/d20180405-30789-un18q8/dmg.DXLEyV"]
==> Creating metadata directory /usr/local/Caskroom/pdftotext/.metadata/latest/20180405203733.447.
==> Creating metadata subdirectory /usr/local/Caskroom/pdftotext/.metadata/latest/20180405203733.447/Casks.
==> Installing artifacts
==> 2 artifact/s defined
#<SortedSet:0x007f80aa1f09a0>
==> Installing artifact of class Hbc::Artifact::Pkg
==> Running installer for pdftotext; your password may be necessary.
==> Package installers may write to any location; options such as --appdir are ignored.
==> Executing: ["/usr/bin/sudo", "-E", "--", "/usr/sbin/installer", "-pkg", "/usr/local/Caskroom/pdftotext/latest/Installer.pkg", "-target", "/", "-verboseR"]
==> installer: Package name is pdftotext
==> installer: Installing at base path /
==> installer:PHASE:Preparing for installation…
==> installer:PHASE:Preparing the disk…
==> installer:PHASE:Preparing pdftotext…
==> installer:PHASE:Waiting for other installations to complete…
==> installer:PHASE:Configuring the installation…
==> installer:STATUS:
==> installer:%2.229250
==> installer:PHASE:Validating packages…
==> installer:%97.750000
==> installer:STATUS:
==> installer:PHASE:Finishing the Installation…
==> installer:STATUS:
==> installer:%100.000000
==> installer:PHASE:The software was successfully installed.
==> installer: The install was successful.
==> Purging files for version latest of Cask pdftotext
🍺  pdftotext was successfully upgraded!
```

One second later, I run it again, and get the same:

```
╰─➤  brew cask upgrade --greedy --verbose --debug
==> Upgrading 3 outdated packages, with result:
dropbox latest, google-backup-and-sync latest, pdftotext latest
==> Started upgrade process for Cask dropbox
==> Hbc::Installer#fetch
==> Satisfying dependencies
==> Downloading
==> Downloading https://www.dropbox.com/download?plat=mac&full=1
/usr/bin/curl --show-error --user-agent Homebrew/1.5.14-3-g94c0d83 (Macintosh; Intel Mac OS X 10.13.3) curl/7.54.0 --fail --location --remote-time --continue-at - --output /Users/reagle/Library/Caches/Homebrew/Cask/dropbox--latest.incomplete https://www.dropbox.com/download?plat=mac&full=1
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0^C
...
```

#### Output of `brew cask doctor`

```
─➤  brew cask doctor                                                                       130 ↵
==> Homebrew-Cask Version
Homebrew-Cask 1.5.14-3-g94c0d83
caskroom/homebrew-cask (git revision f3721; last commit 2018-04-05)
==> macOS
10.13.3
==> SIP
Enabled
==> Java
N/A
==> Homebrew-Cask Install Location
<NONE>
==> Homebrew-Cask Staging Location
/usr/local/Caskroom
==> Homebrew-Cask Cached Downloads
~/Library/Caches/Homebrew/Cask (3 files, 49.9MB)
==> Homebrew-Cask Taps:
/usr/local/Homebrew/Library/Taps/caskroom/homebrew-cask (3956 casks)
==> Contents of $LOAD_PATH
/usr/local/Homebrew/Library/Homebrew/cask/lib
/usr/local/Homebrew/Library/Homebrew
/Library/Ruby/Gems/2.3.0/gems/did_you_mean-1.0.0/lib
/Library/Ruby/Site/2.3.0
/Library/Ruby/Site/2.3.0/x86_64-darwin17
/Library/Ruby/Site/2.3.0/universal-darwin17
/Library/Ruby/Site
/System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/vendor_ruby/2.3.0
/System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/vendor_ruby/2.3.0/x86_64-darwin17
/System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/vendor_ruby/2.3.0/universal-darwin17
/System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/vendor_ruby
/System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/2.3.0
/System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/2.3.0/x86_64-darwin17
/System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/2.3.0/universal-darwin17
==> Environment Variables
LC_ALL="en_US.UTF-8"
PATH="/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/Homebrew/Library/Homebrew/shims/scm"
SHELL="/bin/zsh"
```

#### Output of `brew tap`

```
╰─➤  brew tap
buo/cask-upgrade
caskroom/cask
homebrew/core
homebrew/services
```

**Comment by commitay** [comment]
```
       upgrade [--force] [--greedy] token [ token ... ]
              Without  token  arguments,  upgrade all the installed Casks that
              have newer versions available in the tap; otherwise  update  the
              tokens given in the command line. If --greedy is given then also
              upgrade the Casks having auto_updates true or version :latest.
```

**Comment by reagle** [comment]
@commitay thank you for the response, but I don't understand what that means. Where is that documentation from? Where is the meaning of ":latest" documented? 

My goal is to upgrade any updates *without* re-downloading and installing the same versions, as happens presently. What would the command for that be?

**Comment by commitay** [comment]
> Where is that documentation from?

`man brew-cask`

> Where is the meaning of ":latest" documented

https://github.com/caskroom/homebrew-cask/tree/master/doc/cask_language_reference

> My goal is to upgrade any updates without re-downloading and installing the same versions, as happens presently. What would the command for that be?

`brew cask upgrade` without the `--greedy` option

**Comment by reagle** [comment]
On 4/5/18 5:16 PM, commitay wrote:
> |man brew-cask|

Ah, did not know there was a man page for cask (with a hyphen). I had tried `brew cask help`. Maybe that should say there is a more complete man page?

> https://github.com/caskroom/homebrew-cask/tree/master/doc/cask_language_reference

This doesn't explain what ":latest" means nor why it would continue to pull something it already has.

>  My goal is to upgrade any updates without re-downloading and
> installing the same versions, as happens presently. What would the
> command for that be?
> 
> |brew cask upgrade| without the |--greedy| option

I only resorted to using the greedy option because I found there were lots of packages not being updated. I'll try to document that here once I come across that again.
